src/crypto/crypto.cpp

```diff
@@ -598,4 +598,182 @@ POP_WARNINGS
     sc_sub(&h, &h, &sum);
     return sc_isnonzero(&h) == 0;
   }
+
+  struct rs_comm_borromean_inner {
+    hash h;
+    ec_point L, R;
+    uint64_t i, j;
+  };
+
+  void crypto_ops::generate_borromean_signature(const hash &prefix_hash, const std::vector<key_image> &images, const std::vector<std::vector<public_key>> &pubs,
+    const std::vector<secret_key> &secs, const std::vector<std::size_t> &sec_indices,
+    borromean_signature &sig) {
+    const size_t n = pubs.size();
+    if (images.size() != n) {
+      local_abort("invalid images size");
+    }
+    if (secs.size() != n) {
+      local_abort("invalid secs size");
+    }
+    if (sec_indices.size() != n) {
+      local_abort("invalid sec_indices size");
+    }
+    sig.r.resize(n);
+    for (size_t i = 0; i < n; ++i) {
+      const size_t ring_size = pubs[i].size();
+      assert(sec_indices[i] < ring_size);
+      sig.r[i].resize(ring_size);
+#if !defined(NDEBUG)
+      ge_p3 t;
+      public_key t2;
+      key_image t3;
+      assert(sc_check(&secs[i]) == 0);
+      ge_scalarmult_base(&t, &secs[i]);
+      ge_p3_tobytes(&t2, &t);
+      assert(pubs[i][sec_indices[i]] == t2);
+      generate_key_image(pubs[i][sec_indices[i]], secs[i], t3);
+      assert(images[i] == t3);
+      for (size_t j = 0; j < ring_size; ++j) {
+        assert(check_key(pubs[i][j]));
```

Comment:
I don't think your concern applies here because the above `generate_key_image` uses `pubs[i][sec_indices[i]]` which is ensured to be equal to `t2` which was just now generated from the provided secret key `ge_scalarmult_base(&t, &secs[i]);    ge_p3_tobytes(&t2, &t);`.

After all, I've started making this part by copying the code from the classic `crypto_ops::generate_ring_signature` originating from CryptoNote which seems to have no flaws.